sqlalchemy-stubs/sql/type_api.pyi

```diff
@@ -32,87 +47,85 @@ class TypeEngine(Traversible, Generic[_T]):
+    def adapt(self, __cls: Type[_U], **kw: Any) -> _U: ...
+    def coerce_compared_value(self, op: Any, value: Any) -> TypeEngine[Any]: ...
+    def compile(self, dialect: Optional[Any] = ...) -> Any: ...
 
 class VisitableCheckKWArg(util.EnsureKWArgType, TraversibleType): ...
 
 class UserDefinedType:
     __visit_name__: str = ...
     ensure_kwarg: str = ...
-    def coerce_compared_value(self, op: Any, value: Any): ...
+    def coerce_compared_value(self, op: Any, value: Any) -> Any: ...
 
 class Emulated:
-    def adapt_to_emulated(self, impltype: Any, **kw: Any): ...
-    def adapt(self, impltype: Any, **kw: Any): ...
+    def adapt_to_emulated(self, impltype: Any, **kw: Any) -> Any: ...
+    def adapt(self, __impltype: Any, **kw: Any) -> Any: ...
 
 class NativeForEmulated:
     @classmethod
-    def adapt_native_to_emulated(cls, impl: Any, **kw: Any): ...
+    def adapt_native_to_emulated(cls, impl: Any, **kw: Any) -> Any: ...
     @classmethod
-    def adapt_emulated_to_native(cls, impl: Any, **kw: Any): ...
+    def adapt_emulated_to_native(cls: Type[_NFE], impl: Any, **kw: Any) -> _NFE: ...
 
-_TD = TypeVar("_TD")
-
-class TypeDecorator(SchemaEventTarget, TypeEngine[Any], Generic[_TD]):
+class TypeDecorator(SchemaEventTarget, TypeEngine[_T]):
     __visit_name__: str = ...
     impl: Any = ...
     def __init__(self, *args: Any, **kwargs: Any) -> None: ...
-    coerce_to_is_types: Any = ...
+    coerce_to_is_types: Tuple[Type[Any], ...] = ...
     class Comparator(TypeEngine.Comparator):
         def operate(self, op: Any, *other: Any, **kwargs: Any): ...
         def reverse_operate(self, op: Any, other: Any, **kwargs: Any): ...
     @property
-    def comparator_factory(self): ...
-    def type_engine(self, dialect: Any): ...
-    def load_dialect_impl(self, dialect: Any): ...
-    def __getattr__(self, key: Any): ...
-    def process_literal_param(self, value: Any, dialect: Any) -> None: ...
-    def process_bind_param(self, value: Any, dialect: Any) -> None: ...
-    def process_result_value(self, value: Any, dialect: Any) -> None: ...
-    def literal_processor(self, dialect: Any): ...
-    def bind_processor(self, dialect: Any): ...
-    def result_processor(self, dialect: Any, coltype: Any): ...
-    def bind_expression(self, bindparam: Any): ...
-    def column_expression(self, column: Any): ...
-    def coerce_compared_value(self, op: Any, value: Any): ...
-    def copy(self, **kw: Any): ...
-    def get_dbapi_type(self, dbapi: Any): ...
-    def compare_values(self, x: Any, y: Any): ...
+    def comparator_factory(self) -> Type[Any]: ...  # type: ignore[override]
+    def type_engine(self, dialect: Any) -> TypeEngine[Any]: ...
+    def load_dialect_impl(self, dialect: Any) -> TypeEngine[Any]: ...
+    def __getattr__(self, key: Any) -> Any: ...
+    def process_literal_param(self, value: Any, dialect: Any) -> Optional[str]: ...
+    def process_bind_param(self, value: Any, dialect: Any) -> Optional[Text]: ...
```

Comment:
I don't think the return type has to be a string here and e.g. the documentation provides an example where it's a `datetime` object: https://docs.sqlalchemy.org/en/14/core/custom_types.html#store-timezone-aware-timestamps-as-timezone-naive-utc.

Reply:
I based some of this on `sqlalchemy-stubs` (I wrote most of the typings for `sqlalchemy.sql`), so this is probably a hold-over from those typings. For a lot of this, I tried to figure out what was needed based on how it is used, but that may be wrong (and I wrote those typings over a year ago). I'll look a little more carefully at how this is used and update it accordingly.

Reply:
I updated the typings to (hopefully) be more accurate to the code